lib/datadog/core/configuration/settings.rb

```diff
@@ -646,23 +646,13 @@ def initialize(*_)
           # @default `nil`.
           # @return [Array<String>,nil]
           option :extra_services do |o|
+            o.type :array
+            o.default []
             o.setter do |v|
-              if v.nil?
-                []
-              elsif !v.is_a?(Array)
-                Datadog.logger.warn(
-                  "Unexpected remote.extra_services type #{v.class}, expected Array. " \
-                  'Using default value `[]`'
-                )
-
-                []
-              elsif (t = v.find { |e| !e.is_a?(String) })
-                Datadog.logger.warn(
-                  "Unexpected remote.extra_services element type #{t.class}, expected String. " \
-                  'Using default value `[]`'
-                )
+              if v.any? { |e| !e.is_a?(String) }
+                t = v.find { |e| !e.is_a?(String) }
 
-                []
+                raise ArgumentError, "Unexpected remote.extra_services element type #{t.class}, expected String."
```

Comment:
> we should support at least simple element runtime typing, e.g using o.type 'Array[String]' RBS syntax

That could be an excellent north start for our DSL, but I would only place a little effort now. Having a complex type-checking system is a huge task. For now, the idea behind `o.type :array` without the types inside the array forces us to move the logic to the setter block. Once we see enough use cases for a more sophisticated type check system we could add it. 

> we should attempt to unify to raising instead of logging when items fail the runtime typecheck

The internal type check system already raises https://github.com/DataDog/dd-trace-rb/blob/40778dbaa753510eb37c3322101b0eab49d89965/lib/datadog/core/configuration/option.rb#L214

As to when custom setter blocks fails, I think is up to the implementer to decided